implementation of xdg-foreign protocol

ammen99 opened the pull request:
Supersedes #1841 (for now only rebased) and thus fixes #1702 

TODO:

- [x] Make xdg-foreign v1 and v2 share resources (as discussed in https://github.com/swaywm/wlroots/pull/1841#issuecomment-721125163)

ammen99 commented:
I added a new `wlr_xdg_foreign_registry`. The compositors now can choose whether to let v1 and v2 share handles or not.

A basic implementation in a compositor is here: https://github.com/WayfireWM/wayfire/pull/892

emersion commented on include/util/uuid.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef UTIL_UUID_H
+#define UTIL_UUID_H
+
+int generate_uuid(char out[static 37]);
```
For consistency with the rest of wlroots, we could change this to return a `bool`.

emersion commented on util/uuid.c:
```diff
@@ -0,0 +1,32 @@
+#include "util/uuid.h"
```
Nit: local includes should come after external includes.

emersion commented on util/meson.build:
```diff
@@ -7,3 +7,14 @@ wlr_files += files(
+
+if uuid.found()
+	wlr_deps += uuid
+	add_project_arguments('-DHAS_LIBUUID', language: 'c')
```
We try to avoid `#ifdef`s and use `#if` instead so that forgetting to define the macro triggers an error.

emersion commented on types/meson.build:
```diff
@@ -66,3 +66,11 @@ wlr_files += files(
 	'wlr_xdg_output_v1.c',
 )
+if uuid.found() or uuid_create
+	conf_data.set10('WLR_HAS_XDG_FOREIGN', true)
```
Can we move this one line to the main `meson.build`? It's too easy to incorrectly include another Meson file that depends on this being set before executing `subdir('types')`.

emersion commented on meson.build:
```diff
@@ -107,6 +108,8 @@ libinput = dependency('libinput', version: '>=1.14.0')
 udev = dependency('libudev')
 pixman = dependency('pixman-1')
+uuid = dependency('uuid', required: false)
+uuid_create = cc.has_function('uuid_create')
```
I think we should have a Meson feature option for xdg-foreign, just like we have for all other features. Package maintainers don't want auto-detection of features, they want hard failures when a dep is missing.

ammen99 replied:
I have added this, let me know what you think of the approach.

emersion commented on include/wlr/types/wlr_xdg_foreign_registry.h:
```diff
@@ -0,0 +1,76 @@
+	struct wl_list link; // wlr_xdg_foreign_registry::exported_surfaces
+	struct wlr_xdg_foreign_registry *registry;
+
+	struct wlr_surface *surface;
```
We don't have a destroy listener for this surface. Do we rely on the caller to properly destroy the `wlr_xdg_foreign_exported` when the surface is destoyed?

If so, we should properly document that.

ammen99 replied:
yeah, I explained in in the docstring for init()

emersion commented on include/wlr/types/wlr_xdg_foreign_registry.h:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @return True if the initialization was successful.
+ */
+bool wlr_xdg_foreign_registry_initialize_exported(
```
Maybe rename to `wlr_xdg_foreign_exported_init` for consistency with the rest of wlroots?

ammen99 replied:
I don't like `wlr_xdg_foreign_exported_init` because the registry initializes the given exported surface, it is not like the exported surface initializes itself (if it were C++ I'd put this method inside the registry class). If you however really don't like the name I can rename it, no problem.

emersion replied:
I was thinking of changing the order of arguments too:

```c
bool wlr_xdg_foreign_exported_init(struct wlr_xdg_foreign_exported *surface, struct wlr_xdg_foreign_registry *registry);
```

which would mean "initialize the exported foreign given this registry".

I'm also fine with keeping the current order, but I'd really like to use the `init`/`finish` wording we use elsewhere.

emersion commented on include/wlr/types/wlr_xdg_foreign_registry.h:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Add the given exported surface to the registry and assign it a unique handle.
+ *
+ * @return True if the initialization was successful.
```
We don't use `@return` comments.

emersion commented on include/wlr/types/wlr_xdg_foreign_registry.h:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Remote the given surface from the registry.
+ */
+void wlr_xdg_foreign_registry_remove_exported(
```
Maybe rename to `wlr_xdg_foreign_exported_fini` for consistency with the rest of wlroots?

emersion replied:
Also, we shouldn't need to provide the registry here, `wlr_xdg_foreign_exported` already has a registry field.

emersion commented on include/wlr/types/wlr_xdg_foreign_v2.h:
```diff
@@ -0,0 +1,85 @@
+struct wlr_xdg_foreign_v2 {
+	struct {
+		struct wl_global *global;
+		struct wl_list resources;
```
We don't have a `resources` list for globals anymore.

emersion commented on include/wlr/types/wlr_xdg_foreign_v2.h:
```diff
@@ -0,0 +1,85 @@
+	struct {
+		struct wl_global *global;
+		struct wl_list resources;
+		struct wl_list clients;
```
Do we really need per-client lists? Can't we just have directly a list of `wlr_xdg_exported_v2`/`wlr_xdg_imported_v2`?

ammen99 replied:
yeah the list is actually of wlr_xdg_exported_v2 not of clients, I think `objects` would be a better name, or what do you think?

emersion commented on include/wlr/types/wlr_xdg_foreign_v2.h:
```diff
@@ -0,0 +1,85 @@
+
+struct wlr_xdg_foreign_v2 *wlr_xdg_foreign_v2_create(
+		struct wl_display *display, struct wlr_xdg_foreign_registry *registry);
+void wlr_xdg_foreign_v2_destroy(struct wlr_xdg_foreign_v2 *foreign);
```
We don't expose destructors for globals like this one anymore.

emersion commented on types/wlr_xdg_foreign_registry.c:
```diff
@@ -0,0 +1,77 @@
+#include <stdlib.h>
+#include <string.h>
+
+/**
```
The doc-comment is already in the header file. Would prefer not to duplicate it.

emersion commented on types/wlr_xdg_foreign_registry.c:
```diff
@@ -0,0 +1,77 @@
+
+struct wlr_xdg_foreign_exported *wlr_xdg_foreign_registry_find_by_handle(
+		struct wlr_xdg_foreign_registry *registry, const char *handle) {
+
```
Nit: unnecessary newline.

emersion commented on types/wlr_xdg_foreign_registry.c:
```diff
@@ -0,0 +1,77 @@
+void wlr_xdg_foreign_registry_remove_exported(
+		struct wlr_xdg_foreign_registry *registry,
+		struct wlr_xdg_foreign_exported *surface) {
+
```
Nit: unnecessary newline.

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,472 @@
+	wl_resource_destroy(resource);
+}
+
+static bool verify_is_toplevel(struct wl_resource *client_resource,
```
Hm, why do we need to check whether the surface is a toplevel and send a protocol error if not? The protocol doesn't define any errors like this.

ammen99 replied:
Should we instead silently fail?

emersion replied:
No, reading the protocol again, sending a protocol error is what we should do, but we're just missing the error values.

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/57

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,472 @@
+	struct wl_resource *resource, *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource,
+			&foreign->importer.resources) {
+		wl_resource_destroy(resource);
```
The resources should already be destroyed by the time the `wl_display` is.

emersion reviewed (changes requested):
Thanks for continuing the effort on this feature! Here's a first round of comments.

Apart from these comments, I'd also like the commits to be re-organized and split. For instance the xdg-shell change can be in a standalone commit, so can be the introduction of `util/uuid`. You can use [`Co-authored-by` tags](https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors) in the commit message to indicate a secondary commit author.

ammen99 commented:
Thanks for the review, I'm working on these changes. I will be just pushing new commits until we're ready for merging, and then I'll reorganize the commits as you suggested.

ammen99 commented:
I think I have addressed all of your comments, except the one about erroring when the handle is not a toplevel. The protocol says `Only xdg_surface surfaces may be exported.` which I think is enough grounds to disconnect a client which exports something else.

emersion commented on include/util/uuid.h:
```diff
@@ -0,0 +1,8 @@
+
+#include <stdbool.h>
+
+bool generate_uuid(char out[static 37]);
```
Maybe we should have a constant for this?

```c
#ifndef UUID_STR_LEN
#define UUID_STR_LEN 37
#endif
```

emersion replied:
Hm, I guess it wouldn't help too much, since this is a private header.

emersion commented on meson.build:
```diff
@@ -110,6 +111,17 @@ pixman = dependency('pixman-1')
+	uuid = dependency('uuid', required: false)
+	uuid_create = cc.has_function('uuid_create')
+	if uuid.found() or uuid_create
+		conf_data.set10('WLR_HAS_XDG_FOREIGN', 1)
```
Nit: `set10` takes a `bool`, so should be `true` instead of 1.

emersion commented on meson.build:
```diff
@@ -110,6 +111,17 @@ pixman = dependency('pixman-1')
+	if uuid.found() or uuid_create
+		conf_data.set10('WLR_HAS_XDG_FOREIGN', 1)
+	elif get_option('xdg-foreign').enabled()
+		error('Missing dependency uuid and uuid_create function not available' +
```
Nit: missing space in error message.

emersion commented on util/meson.build:
```diff
@@ -7,3 +7,16 @@ wlr_files += files(
+		add_project_arguments('-DHAS_LIBUUID=0', language: 'c')
+	endif
+	wlr_files += files(
+			'uuid.c',
```
Nit: this can just be single-line

emersion commented on util/uuid.c:
```diff
@@ -0,0 +1,31 @@
+	if (status != uuid_s_ok) {
+		return false;
+	}
+	strcpy(out, str);
```
`strcpy` is insecure. Use `memcpy` instead. Maybe `assert(strlen(str) + 1 == 37)` would be nice too.

emersion commented on include/wlr/types/wlr_xdg_foreign_v1.h:
```diff
@@ -0,0 +1,86 @@
+	struct wl_global *importer_global;
+
+	struct {
+		struct wl_global *global;
```
We have `exporter_global` and `exporter.global`. Do we need both?

ammen99 replied:
Indeed, this is a mistake, will remove.

emersion commented on types/wlr_xdg_foreign_v1.c:
```diff
@@ -0,0 +1,458 @@
+		return;
+	}
+
+	wlr_signal_emit_safe(&foreign->events.destroy, foreign);
```
Use NULL as data argument to allow future extensibility.

emersion commented on types/wlr_xdg_foreign_v1.c:
```diff
@@ -0,0 +1,458 @@
+
+	wl_list_insert(&exporter->exports, &exported->link);
+	wl_list_init(&exported->imports);
+	wl_signal_init(&exported->base.events.destroy);
```
This is already done by `wlr_xdg_foreign_exported_init`

ammen99 replied:
It isn't done but probably should be :)

emersion commented on types/wlr_xdg_foreign_v1.c:
```diff
@@ -0,0 +1,458 @@
+	struct wlr_xdg_exporter_v1 *exporter = xdg_exporter_from_resource(resource);
+	struct wlr_xdg_exported_v1 *exported, *exported_tmp;
+	wl_list_for_each_safe(exported, exported_tmp, &exporter->exports, link) {
+		wl_resource_destroy(exported->resource);
```
Nothing in the protocol says that destroying the exporter also destroys the child objects.

emersion commented on types/wlr_xdg_foreign_v1.c:
```diff
@@ -0,0 +1,458 @@
+		xdg_importer_from_resource(resource);
+	struct wlr_xdg_imported_v1 *imported, *imported_tmp;
+	wl_list_for_each_safe(imported, imported_tmp, &importer->imports, link) {
+		disconnect_imported(imported);
```
Nothing in the protocol says that destroying the importer also destroys the child objects.

emersion commented on include/wlr/types/wlr_xdg_foreign_v1.h:
```diff
@@ -0,0 +1,86 @@
+	void *data;
+};
+
+struct wlr_xdg_exporter_v1 {
```
I still don't think we need this intermediate struct. We can just store the imported/exported objects in the `wlr_xdg_foreign_v1`.

ammen99 replied:
You are right! I misunderstood your previous comment. This change removed a lot of unnecessary code.

ammen99 commented:
Alright, made it so that we don't store anything about the intermediate exporter/importer resources. They directly have the `foreign` as their user data, which is enough.

emersion commented on types/wlr_xdg_foreign_registry.c:
```diff
@@ -0,0 +1,72 @@
+	return NULL;
+}
+
+void wlr_xdg_foreign_exported_finish(struct wlr_xdg_foreign_exported *surface) {
```
Maybe we should trigger the `destroy` event here?

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+
+	imported->exported = wlr_xdg_foreign_registry_find_by_handle(
+		foreign->registry, handle);
+	imported->resource = wl_resource_create(wl_client,
```
Missing a NULL check here

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+	wl_list_insert(&foreign->importer.objects, &imported->link);
+
+	if (imported->exported == NULL) {
+		wl_list_init(&imported->exported_destroyed.link);
```
I'd prefer to make imported objects without an exported inert, by making the user data pointer NULL and free'ing the `struct wlr_xdg_imported_v2`. This is the standard thing that wlroots does when objects are still alive but can't be used anymore.

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+	}
+
+	exported->base.surface = surface;
+	exported->resource = wl_resource_create(wl_client,
```
Missing a NULL check here.

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+	zxdg_exported_v2_send_handle(exported->resource, exported->base.handle);
+
+	exported->xdg_surface_unmap.notify = handle_xdg_surface_unmap;
+	if (wlr_surface_is_xdg_surface(surface)) {
```
We checked `verify_is_toplevel` above, so this should be unnecessary?

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_exported_v2 *exported =
+		wl_container_of(listener, exported, xdg_surface_unmap);
+	wl_resource_destroy(exported->resource);
```
This destroys the resource behind the client's back. We can't do this, otherwise the client will die when sending an `xdg_exported.destroy` request.

Why do we need to handle `unmap` in the first place? The protocol doesn't seem to mention anything about it.

ammen99 replied:
A good question. I will remove that as well.

ammen99 replied:
Hmm, I suppose we want to listen not for unmap, but for the case where the toplevel is destroyed.

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+		return;
+	}
+	if (wlr_surface_is_xdg_surface(wlr_surface)
+			!= wlr_surface_is_xdg_surface(wlr_surface_child)) {
```
Hm, `verify_is_toplevel` should take care of this too I think?

ammen99 replied:
Yes, it was a leftover from the time when there was both xdg6 and xdg-stable

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+	child->xdg_surface_unmap.notify = handle_child_xdg_surface_unmap;
+	child->xdg_toplevel_set_parent.notify = handle_xdg_toplevel_set_parent;
+
+	if (wlr_surface_is_xdg_surface(wlr_surface)) {
```
This should always be true as well I think.

emersion commented on types/wlr_xdg_foreign_v2.c:
```diff
@@ -0,0 +1,411 @@
+				&child->xdg_surface_unmap);
+		wl_signal_add(&surface_child->toplevel->events.set_parent,
+				&child->xdg_toplevel_set_parent);
+	}
```
`child` is leaked in the else branch, because `wlr_xdg_toplevel_set_parent` won't free it in `xdg_imported_handle_resource_destroy`.

ammen99 replied:
if we remove the if() there is no else, right?

ammen99 commented:
I think I have addressed your new review comments. Inert state is used in two different places:

- For imported, if the exported could not be found
- For exported, if the xdg-surface is destroyed (not unmapped, as it was previously).

emersion reviewed (commented):
This looks good to me, thanks. Can you split everything into logical commits? See https://github.com/swaywm/wlroots/pull/2487#pullrequestreview-546948373

ammen99 commented:
> This looks good to me, thanks. Can you split everything into logical commits? See [#2487 (review)](https://github.com/swaywm/wlroots/pull/2487#pullrequestreview-546948373)

Thanks for the review! I have split everything into five commits.

emersion reviewed (approved):
Nice work, thanks for you contribution!
